include/emp/prefab/ControlPanel.hpp

```diff
@@ -0,0 +1,331 @@
+       */
+      void SetRate(const int & rate) {
+        refresh_rates[refresh_unit] = rate;
+      }
+
+      /**
+       * Get the refresh list for this control panel
+       * @return a list of Widgets that will be refreshed every update period
+       */
+      emp::vector<web::Widget> & GetRefreshList() {
+        return refresh_list;
+      }
+    };
+  }
+  /**
+   * Use the ConfigPanel class to add a play/pause toggle button and a step
+   * button to your application. You can add a simulation to be run, web
+   * components to be redrawn, and more Buttons or ButtonGroups to add more
+   * functionality to the control panel.
+   */
+  class ControlPanel : public web::Div {
+
+    /**
+     * Get shared info pointer, cast to ControlPanel-specific type.
+     * @return cast pointer
+     */
+    internal::ControlPanelInfo * Info() {
+      return dynamic_cast<internal::ControlPanelInfo *>(info);
+    }
+
+    /**
+     * Get shared info pointer, cast to const ControlPanel-specific type.
+     * @return cast pointer
+     */
+    const internal::ControlPanelInfo * Info() const {
+      return dynamic_cast<internal::ControlPanelInfo *>(info);
+    }
+
+    ToggleButtonGroup toggle_run;
+    Div button_line;
+    web::Button step;
+
+    protected:
+    /**
+     * The protected contructor for a Control panel that sets up the state
+     * and event handlers
+     * @param refresh_mode units of "MILLISECONDS" or "FRAMES"
+     * @param refresh_rate the number of milliseconds or frames between refreshes
+     * @param in_info info object associated with this component
+     */
+    ControlPanel(
+      const int & refresh_rate,
+      const std::string & refresh_unit,
+      web::internal::DivInfo * in_info
+    ) : web::Div(in_info),
+    toggle_run{
+      FontAwesomeIcon{"fa-play"}, FontAwesomeIcon{"fa-pause"},
+      "success", "warning",
+      true, false,
+      emp::to_string(GetID(), "_main_toggle")
+    },
+    button_line(ButtonGroup{emp::to_string(GetID(), "_main")}),
+    step{
+      [](){ ; },
+      "<span class=\"fa fa-step-forward\" aria-hidden=\"true\"></span>",
+      emp::to_string(GetID(), "_main_step")
+    }
+    {
+      AddAttr(
+        "class", "btn-toolbar",
+        "class", "space_groups",
+        "role", "toolbar",
+        "aria-label", "Toolbar with simulation controls"
+      );
+      SetRefreshRate(refresh_rate, refresh_unit);
+      step.AddAttr(
+        "class", "btn",
+        "class", "btn-success"
+      );
+
+      static_cast<Div>(*this) << button_line;
+      button_line << toggle_run;
+      button_line << step;
+
+      AddAnimation(GetID(),
+        [&run_sim=GetSimulation(),
+          &refresh_list=Info()->GetRefreshList(),
+          &do_redraw=Info()->GetRedrawChecker()]
+        (const web::Animate & anim) mutable {
+          // Run the simulation function every frame
+          run_sim();
+          // Redraw widgets according to a rule
+          if(do_redraw(anim)) {
```

Comment:
rename do_redraw should_redraw

I might also move auto& wid : refresh_list wid.Redraw to a helper method (probably private?) called do_redraw

Reply:
or DoRedraw